### PR TITLE
RND-168 Dell Icons - part 2 - export ts object

### PR DESCRIPTION
**Pull request by leon-good-life**
## Description
follow up to https://github.com/cloudify-cosmo/cloudify-ui-common/pull/157

PR in stage: https://github.com/cloudify-cosmo/cloudify-stage/pull/2494

## Screenshots / Videos
Screenshot 2023-04-27 at 13 21 07

## Checklist

- [x] My code follows the [UI Code Style](https://cloudifysource.atlassian.net/l/c/x8fq902p).
- [x] I followed the [UI Testing Guidelines](https://cloudifysource.atlassian.net/l/cp/udCVfvrx) and verified that all tests/checks have passed.
- [x] I verified if that change requires any update in the [documentation](https://cloudifysource.atlassian.net/l/c/4XKtPocy).
- [x] I added proper labels to this PR.

## Tests
N/A

 

## Documentation
Will be done at the same time of finalising the PR in `cloudify-tage`.

### CODE REVIEW

File: packages/frontend/styles/ddsIcons.ts [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+    'wrench-manage',
+    'zoom-in',
+    'zoom-out'
+];
+
+const iconClassName = (iconSuffix: string) => `icon-button dds__icon dds__icon--${iconSuffix}`;
+
+const dashToCamelCase = (dashCase: string) => dashCase.replace(/-([a-z])/g, g => g[1].toUpperCase());
+
+export const icons: Record<typeof iconsSuffix[number], string> = iconsSuffix.reduce(
```
Comment (Vorbert-Kruk) [inline packages/frontend/styles/ddsIcons.ts:464]: As `typeof iconsSuffix[number]` is now returning `string`, we can simplify code, by using it directly 💪 

```suggestion
export const icons: Record<string, string> = iconsSuffix.reduce(
```

I suspected that the thought behind the implementation was to list all available icon names, which could be accessed from `icons` object, but to make this solution work, some structural changes would need to happen

File: packages/frontend/styles/ddsIcons.ts [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+    'view-table',
+    'virtualization',
+    'wrench-manage',
+    'zoom-in',
+    'zoom-out'
+];
+
+const iconClassName = (iconSuffix: string) => `icon-button dds__icon dds__icon--${iconSuffix}`;
+
+const dashToCamelCase = (dashCase: string) => dashCase.replace(/-([a-z])/g, g => g[1].toUpperCase());
```
[the remark this answers is not in the document]
Comment (leon-good-life) [reply packages/frontend/styles/ddsIcons.ts]: I refactored it: [0c70ee7](https://github.com/cloudify-cosmo/cloudify-ui-common/pull/164/commits/0c70ee77dc435f54e97e184485b05448b094361a)

File: packages/frontend/styles/ddsIcons.ts [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+    'view-list',
+    'view-row',
+    'view-table',
+    'virtualization',
+    'wrench-manage',
+    'zoom-in',
+    'zoom-out'
+];
+
+const iconClassName = (iconSuffix: string) => `icon-button dds__icon dds__icon--${iconSuffix}`;
```
[the remark this answers is not in the document]
Comment (leon-good-life) [reply packages/frontend/styles/ddsIcons.ts]: I removed it, and I added a function I export that adds `icon-button`: [1966d77](https://github.com/cloudify-cosmo/cloudify-ui-common/pull/164/commits/1966d773fe91261185c76e7cbf31890ce91e5bf7)

**Comment by qooban** [comment]
Sorry, I have to opt out as I'm leaving for 1-week vacation. Please work with @kubama . 

I guess this PR can help in [RND-563](https://cloudifysource.atlassian.net/browse/RND-563).

File: packages/frontend/styles/ddsIcons.ts [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+
+export const icons: Record<string, string> = iconsSuffix.reduce(
+    (iconsObj, iconSuffix) => ({
+        ...iconsObj,
+        [dashToCamelCase(iconSuffix)]: iconClassName(iconSuffix)
+    }),
+    {}
+);
+
+export const iconButtonClassName = (ddsIcon: string) => `icon-button ${ddsIcon}`
```
Comment (Vorbert-Kruk) [inline packages/frontend/styles/ddsIcons.ts:472]: WDYT About the following modification?

```suggestion
export const iconNames = [
'accessibility',
'addSqr',
'airplane',
'alarmBell',
'alarmClock',
'alertCheckCir',
'alertCheckSqr',
'alertError',
'alertInfoCir',
'alertNotice',
'appWindow',
'arrowCirDown',
'arrowCirLeft',
'arrowCirRight',
'arrowCirUp',
'arrowDown',
'arrowExport',
'arrowLeft',
'arrowRight',
'arrowTriDown',
'arrowTriLeft',
'arrowTriRight',
'arrowTriSolidRight',
'arrowTriUp',
'arrowUp',
'assistance',
'audioSpeaker',
'audioSpeakerMute',
'audioSpeakerOn',
'awardCertificate',
'awardTrophy',
'batteryCharge',
'bolt',
'bookmark',
'bookOpen',
'brightness',
'calculator',
'calendar',
'camera',
'cancelled',
'captureImage',
'cardCredit',
'cardInfo',
'carryCase',
'cart',
'chainLink',
'chargeBack',
'chartBarsAxis',
'chartBarsBox',
'chartInsights',
'chatBot',
'chatDots',
'check',
'chevronCirDown',
'chevronCirLeft',
'chevronCirRight',
'chevronCirUp',
'chevronDown',
'chevronLeft',
'chevronRight',
'chevronUp',
'chipSet',
'clipboard',
'clock',
'clockArrowLeft',
'clockArrowRight',
'clone',
'closeX',
'closeBack',
'closeCir',
'closeSqr',
'cloud',
'cloudCheck',
'collapseDownSqr',
'collapseUpSqr',
'comment',
'compare',
'conversation',
'conversationCir',
'copy',
'copyAlt',
'cube',
'currency',
'cursor',
'dDisplay',
'dashboard',
'dataCenter',
'dataCompute',
'datastore',
'debug',
'deviceDataCenter',
'deviceLaptop',
'deviceMobile',
'deviceModular',
'deviceMonitor',
'deviceServer',
'deviceServerTower',
'diagnostic',
'dimensionsWeight',
'discSoftware',
'docAdd',
'docArrowLeft',
'docArrowRight',
'docLines',
'docReports',
'document',
'downloadAlt',
'ear',
'energyEfficient',
'escalateTri',
'exitDoor',
'eyeViewOn',
'fan',
'filter',
'fingerprint',
'flag',
'flame',
'flashDrive',
'folder',
'folderAdd',
'folderArrowRight',
'fullScreen',
'gear',
'globeSimple',
'handle',
'handTouch',
'hardDrive',
'heart',
'heartFilled',
'helpCir',
'hinge',
'home',
'import',
'importAlt',
'inbox',
'inkToner',
'key',
'keyboard',
'keyboardMouse',
'laptopGeneric',
'lightbulb',
'location',
'lockClosed',
'lockOpen',
'logIn',
'mail',
'mailOpen',
'menuClosed',
'microphoneOff',
'minusMinimize',
'missingImage',
'mouse',
'mouseCord',
'movieClapper',
'movieClapperPlay',
'musicNote',
'networkVert',
'office',
'overflow',
'paintBrush',
'paintPalette',
'paperClip',
'partners',
'pause',
'pencil',
'pencilCheck',
'pencilMark',
'performance',
'phone',
'pin',
'playCir',
'popoutWindow',
'popUpArrow',
'powerSupply',
'printer',
'projector',
'recycle',
'recycleBin',
'redo',
'redoMove',
'refreshSpin',
'removeCir',
'removeSqr',
'reset',
'restore',
'save',
'saveDisk',
'scale',
'search',
'searchArrowLeft',
'searchBack',
'searchForward',
'share',
'shield',
'shieldCheck',
'sliderVert',
'snapshot',
'snowflake',
'sortHoriz',
'sortVert',
'star',
'starFilled',
'stop',
'syncHoriz',
'tag',
'thumbDown',
'thumbUp',
'trash',
'tv',
'undo',
'undoMove',
'upload',
'user',
'userGroup',
'userGroups',
'viewColumn',
'viewGrid',
'viewGridAlt',
'warranty',
'weightBell',
'wifi',
'wrenchSingle',
'wrenchTools',
'wifiNone',
'globeLocation',
'brandSkype',
'giftCard',
'metadata',
'objects',
'playSqr',
'plusAdd',
'sortAz',
'ellipsis',
'arrange',
'volume',
'arrowTriSolidStacked',
'popoutDoubleCondenseVertical',
'popoutDoubleCondense',
'barbell',
'expand',
'currencyCoinsPounds',
'currencyCoinsRupees',
'currencyCoinsYen',
'audioCard',
'popoutDoubleVertical',
'audioSpeakerLevels',
'deviceDesktop',
'audioSpeakerOff',
'calendarAdd',
'calendarCheck',
'calendarRemove',
'heartHalfFilled',
'batteryEmpty',
'starHalfFilled',
'thumbDownFilled',
'thumbUpFilled',
'brandAlienware',
'brandDell',
'brandDellemc',
'brandDelltechnologies',
'brandDelltechnologiesvert',
'bookmarkFilled',
'cardSwipeLeft',
'cardSwipeRight',
'chargeBackFolder',
'chargeBackReport',
'clipboardCheck',
'cloudDocument',
'cloudIq',
'columnAdd',
'commentAdd',
'commentCheck',
'deploy',
'deviceLaptopDetachable',
'deviceLockScreen',
'deviceStorageArray',
'deviceSwitchBlade',
'deviceSwitchBlade-2U',
'deviceTabletHoriz',
'deviceTabletVert',
'deviceThinClient',
'discArrowDown',
'discArrowUp',
'brandDellGaming',
'tagDeal',
'alertInfoSqr',
'userDirectory',
'brandWindows',
'appWindowSearch',
'currencyCoinsIndianRupees',
'currencyCoinsGeneric',
'currencyCoinsEuros',
'addCir',
'gameController',
'gamingDesktop',
'batteryHorizEmpty',
'batteryHorizFull',
'batteryHorizHalf',
'batteryVertFull',
'batteryVertHalf',
'bluetooth',
'calendarSearch',
'checkProperty',
'clipboardLines',
'clipboardNotes',
'clipboardRemove',
'clipboardSearch',
'cloudRestore',
'cloudUpload',
'cloudWireless',
'collapseDown',
'collapseUp',
'currencyCoins',
'cursorClock',
'deviceNodes',
'deviceScreenSize',
'discCheck',
'discInsert',
'discRecovery',
'discSecure',
'discSystem',
'docCheck',
'docCube',
'docManual',
'docsCopyAlt',
'docSearch',
'docSecure',
'docSettings',
'docSupport',
'download',
'eyeViewOff',
'filterAdd',
'filterClear',
'filterMove',
'filterRemove',
'folderArrowUp',
'folderMedia',
'gearArrow',
'gearWrench',
'graphicsCard',
'handTouchScreen',
'hdGeneric',
'headphones',
'headset',
'heat',
'helpSqr',
'hostServer',
'hostServerHex',
'hostServerRack',
'inboxArrowDown',
'infrastructure',
'keypad',
'lcd',
'locationPlay',
'logOut',
'leaf',
'manageInfo',
'managementSoftware',
'memory',
'memoryCardReader',
'memorySdCard',
'microchip',
'microphoneOn',
'migrate',
'minimize',
'2In-1',
'preSeed',
'receipt',
'brandXps',
'dataProtection',
'hyperConvergedInfrastructure',
'unpin',
'networkConnected',
'networkHoriz',
'networkShare',
'networkVertDisable',
'partnership',
'password',
'pauseSqr',
'pencilRemove',
'popoutDouble',
'popoutShare',
'popoutWindowArrow',
'popUpArrowCorner',
'portDouble',
'powerOff',
'powerOn',
'protractor',
'pulse',
'puzzle',
'qos',
'rack',
'rss',
'searchArrowRight',
'serverRackArrowRight',
'shipping',
'skipForward',
'skipToEnd',
'skipToFront',
'sliderHoriz',
'smartSelect',
'socialNetworking',
'startUp',
'suspend',
'syncVert',
'taskCheck',
'taskCheckProgress',
'usb',
'userAbout',
'userAdd',
'userCheck',
'userCir',
'userCirAdd',
'userGroupDisable',
'userGroupsDisable',
'userLock',
'userRemove',
'userSearch',
'userUnlock',
'videoCard',
'viewDetailed',
'webCamera',
'webCameraHd',
'wifiSimple',
'windowsSupport',
'wrenchCir',
'wrenchDouble',
'packageClosed',
'packageOpen',
'pieChart',
'pieChartSolid',
'powerSupplyCord',
'powerSupplyDouble',
'processor',
'productivity',
'removableStorageUsb',
'serverRackSearch',
'serverRackSupport',
'shieldProtect',
'sortZa',
'speakerDouble',
'stack',
'toolbox',
'touchPad',
'translate',
'userCirRemove',
'userContact',
'userDisable',
'userGroupChat',
'viewList',
'viewRow',
'viewTable',
'virtualization',
'wrenchManage',
'zoomIn',
'zoomOut'
] as const;

const iconNameToIconClassName = (iconName: string) => iconName.replace(/[A-Z]/g, upperCaseLetter => `-${upperCaseLetter.toLowerCase()}`);

const iconClassName = (iconName: string) => {
    const iconClassName = iconNameToIconClassName(iconName);
    return `dds__icon dds__icon--${iconClassName}`;
};

export const iconButtonClassName = (iconName: string) => {
    const iconClassName = iconNameToIconClassName(iconName);
    return `icon-button ${iconClassName}`;
}

type Icons = Record<typeof iconNames[number], string>;

export const icons: Icons = iconNames.reduce(
    (iconsObj, iconName) => ({
        ...iconsObj,
        [iconName]: iconClassName(iconName)
    }),
    {} as Icons
);

```

It's modifying the idea behind this code by 180 degrees, but it's giving us the ability to have  a typing suggestions regarding icon names
In addition, we could omit typing icon suffixes by hand, in favor of having a dedicated object, which we could use to implement desired icon 🎉 

With previous approach, we'd have to do 
```typescript
<i className={icons['audio-speaker-mute']} />
```

But currently we should be able to do
```typescript
<i className={icons[iconNames.audioSpeakerMute]} />
```

which should be less error prone and more friendly to developers 🎉 

Additional benefit is that we are having a typescript support under the hood, so making
```typescript
<i className={icons['some-random-text']} />
```
will result in getting an error - which wasn't the case with the previous approach

I know that approach modification comes with the great amount of modified lines of code
That's why I thought that I'll make a ready to implement suggestion, as I can see a lot of added value coming out of the presented approach 💪 

Let me know what do you think about it 😉

Comment (leon-good-life) [reply packages/frontend/styles/ddsIcons.ts]: [ec72503](https://github.com/cloudify-cosmo/cloudify-ui-common/pull/164/commits/ec72503cfec39642f0602fec8147986a7f14235a)

**Review by kubama — COMMENTED** [review]
I don't think this PR is the way to go, and as such I'd rather reject it, at least for now.
Please have a look at https://github.com/cloudify-cosmo/cloudify-ui-components/tree/RND-463 - we are going to wrap the whole dds-core library, which contains already all we need (see https://github.com/cloudify-cosmo/cloudify-ui-components/blob/a0468e445ac00dcb7ccea53c22089d95f276c0e9/dds-core/src/components/icon/ts/icon.ts). I don't see a reason to duplicate things.
To keep things simple, I think all icon related stuff should be exposed as part of yet to come dds `Icon` react component.
As a side node - looks like the same area is now being covered by different squads, and there is no communication or agreement how to approach things. To bad.

**Pull request closed without merging**